staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -437,35 +448,86 @@ func (p *patcher) applyPatch(_ context.Context, _, currentObject runtime.Object)
+	admissionOperation := p.operation.admissionType()
+	userInfo, _ := request.UserFrom(ctx)
+	return admission.NewAttributesRecord(updatedObject, currentObject, p.kind, p.namespace, p.name, p.resource, p.subresource, admissionOperation, userInfo)
+}
+
 // applyAdmission is called every time GuaranteedUpdate asks for the updated object,
 // and is given the currently persisted object and the patched object as input.
 func (p *patcher) applyAdmission(ctx context.Context, patchedObject runtime.Object, currentObject runtime.Object) (runtime.Object, error) {
 	p.trace.Step("About to check admission control")
-	return patchedObject, p.admissionCheck(patchedObject, currentObject)
+	mutatingAdmission, ok := p.admissionCheck.(admission.MutationInterface)
+	if ok && mutatingAdmission.Handles(p.operation.admissionType()) {
+		attributes := p.admissionAttributes(ctx, patchedObject, currentObject)
+		return patchedObject, mutatingAdmission.Admit(attributes)
+	}
+	return patchedObject, nil
 }
 
 // patchResource divides PatchResource for easier unit testing
 func (p *patcher) patchResource(ctx context.Context, scope RequestScope) (runtime.Object, error) {
 	p.namespace = request.NamespaceValue(ctx)
 	switch p.patchType {
 	case types.JSONPatchType, types.MergePatchType:
-		p.mechanism = &jsonPatcher{patcher: p}
+		mechanism := &jsonPatcher{patcher: p}
+		p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
 	case types.StrategicMergePatchType:
 		schemaReferenceObj, err := p.unsafeConvertor.ConvertToVersion(p.restPatcher.New(), p.kind.GroupVersion())
 		if err != nil {
 			return nil, err
 		}
-		p.mechanism = &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		mechanism := &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
 	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
 	case types.ApplyPatchType:
-		p.mechanism = &applyPatcher{patcher: p, model: scope.OpenAPISchema}
+		_, err := p.restPatcher.Get(ctx, p.name, &metav1.GetOptions{})
+		if errors.IsNotFound(err) && p.restApplier != nil {
+			// Only attempt to create if the object doesn't exist and the storage implements Applier
+			// since Patcher is not sufficient to allow the Create operation
+			p.operation = &patchCreater{patcher: p}
+		} else {
+			mechanism := &applyPatcher{patcher: p, model: scope.OpenAPISchema}
+			p.operation = &patchUpdater{patcher: p, mechanism: mechanism}
+		}
 	default:
 		return nil, fmt.Errorf("%v: unimplemented patch type", p.patchType)
 	}
-	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyAdmission)
 	return finishRequest(p.timeout, func() (runtime.Object, error) {
-		updateObject, _, updateErr := p.restPatcher.Update(ctx, p.name, p.updatedObjectInfo, p.createValidation, p.updateValidation)
-		return updateObject, updateErr
+		return p.operation.execute(ctx)
```

Comment:
if this can now create, I think we need the status code to reflect that like update does